### DISCUSSION TITLE
v3.0.0-alpha

### DISCUSSION THREAD

**Discussion by lee-to** [moonshine-software/moonshine#1255] [Announcements]
## What's Changed
* [3.x] feat: New layout concept by @lee-to in https://github.com/moonshine-software/moonshine/pull/867
* New menu manager by @lee-to in https://github.com/moonshine-software/moonshine/pull/901
* There are only components around by @lee-to in https://github.com/moonshine-software/moonshine/pull/914
* SDUI by @lee-to in https://github.com/moonshine-software/moonshine/pull/918
* New url generator by @lee-to in https://github.com/moonshine-software/moonshine/pull/924
* New icon system by @lee-to in https://github.com/moonshine-software/moonshine/pull/926
* feat: Applies by @lee-to in https://github.com/moonshine-software/moonshine/pull/928
* Moonshine configurator by @lee-to in https://github.com/moonshine-software/moonshine/pull/942
* refactor: Handlers 3.0 by @lee-to in https://github.com/moonshine-software/moonshine/pull/943
* Api by @lee-to in https://github.com/moonshine-software/moonshine/pull/976
* feat: Resource separate approach is default by @lee-to in https://github.com/moonshine-software/moonshine/pull/989
* fix: publish favicons by @DissNik in https://github.com/moonshine-software/moonshine/pull/1003
* New structure by @lee-to in https://github.com/moonshine-software/moonshine/pull/1005
* Refactoring IndexPage, ActionGroup and Flex, new component  Div by @DissNik in https://github.com/moonshine-software/moonshine/pull/1017
* Popover and UpdateOnPreview in Popover by @lee-to in https://github.com/moonshine-software/moonshine/pull/1029
* Table builder by @lee-to in https://github.com/moonshine-software/moonshine/pull/1034
* refactor: OnlyLink by @lee-to in https://github.com/moonshine-software/moonshine/pull/1050
* Naming by @lee-to in https://github.com/moonshine-software/moonshine/pull/1056
* [3.x] fix dropdown menu position by @DissNik in https://github.com/moonshine-software/moonshine/pull/1062
* [3.x] fixing style close button in the minimalistic theme by @DissNik in https://github.com/moonshine-software/moonshine/pull/1064
* fix(relation): asyncSearch fix by @dev-lnk in https://github.com/moonshine-software/moonshine/pull/1070
* refactor: WithFields by @lee-to in https://github.com/moonshine-software/moonshine/pull/1072
* [3.x] refactoring: remove TinyMce by @DissNik in https://github.com/moonshine-software/moonshine/pull/1092
* [3.x] style for mobilebar when :has not supported by @DissNik in https://github.com/moonshine-software/moonshine/pull/1096
* feat: DI concept by @lee-to in https://github.com/moonshine-software/moonshine/pull/1106
* [3.x] namespace MoonShineConfigurator in the stub by @DissNik in https://github.com/moonshine-software/moonshine/pull/1110
* fix(menu): active item by @dev-lnk in https://github.com/moonshine-software/moonshine/pull/1113
* feat: New buttons and activeActions approach by @lee-to in https://github.com/moonshine-software/moonshine/pull/1121
* refactor: ListOf by @lee-to in https://github.com/moonshine-software/moonshine/pull/1122
* [3.x] top menu by @DissNik in https://github.com/moonshine-software/moonshine/pull/1129
* [3.x] hide mobile sidebar menu when has top menu by @DissNik in https://github.com/moonshine-software/moonshine/pull/1140
* [3.x] refactor: ExportImport sep by @lee-to in https://github.com/moonshine-software/moonshine/pull/1132
* Remove Import/Export by @lee-to in https://github.com/moonshine-software/moonshine/pull/1154
* feat: New Notification system by @lee-to in https://github.com/moonshine-software/moonshine/pull/1156
* refactor: Remove Td, optimize prepareFields, fields to table attr by @lee-to in https://github.com/moonshine-software/moonshine/pull/1166
* [3.x] ShowWhen refactor by @dev-lnk in https://github.com/moonshine-software/moonshine/pull/1152
* feat: Routing where ResourceItem is parameter by @lee-to in https://github.com/moonshine-software/moonshine/pull/1175
* feat: ResourceWithTableModifiers by @lee-to in https://github.com/moonshine-software/moonshine/pull/1189
* feat: TableBuilder sugar by @lee-to in https://github.com/moonshine-software/moonshine/pull/1190
* refactor: NotificationItemContract by @dev-lnk in https://github.com/moonshine-software/moonshine/pull/1203
* feat: CrudResource by @lee-to in https://github.com/moonshine-software/moonshine/pull/1221
* feat: CursorPaginate mode by @lee-to in https://github.com/moonshine-software/moonshine/pull/1223
* feat: Tabs with slots by @lee-to in https://github.com/moonshine-software/moonshine/pull/1224
* chore: input-wrapper -> wrapper by @lee-to in https://github.com/moonshine-software/moonshine/pull/1233
* feat: HasMany top buttons by @lee-to in https://github.com/moonshine-software/moonshine/pull/1238
* fix: When component by @dev-lnk in https://github.com/moonshine-software/moonshine/pull/1246
* [3.x] fix: display logo in the collapse menu by @DissNik in https://github.com/moonshine-software/moonshine/pull/1247
* chore: Monorepo by @lee-to in https://github.com/moonshine-software/moonshine/pull/1253

**Full Changelog**: https://github.com/moonshine-software/moonshine/compare/2.14.4...3.0.0-alpha

<hr /><em>This discussion was created from the release <a href='https://github.com/moonshine-software/moonshine/releases/tag/3.0.0-alpha'>v3.0.0-alpha</a>.</em>

**Comment by ArtemYurov** [comment]
Подскажите как теперь настраивать экспорт\импорт у ресурсов?
Какой логикой заменяется`useOnImport()` ?

**Reply by lee-to** [reply]
> Подскажите как теперь настраивать экспорт\импорт у ресурсов?
> 
> Какой логикой заменяется`useOnImport()` ?

Теперь импорт экспорт в отдельном пакете moonshine-software/import-export

**Comment by ArtemYurov** [comment]
Я правильно понимаю что в ресурсе теперь такого нельзя стандартную страницу модифицировать как тут:
```
    public function pages(): array
    {
        return [
            IndexPage::make($this->title())
                ->pushToLayer(Layer::TOP,
                Collapse::make('Фильтры', [
                        (new FiltersForm())($this)
                    ])->open(true)
                ),
                ....
        ];
    }
```

и нужно будет обязательно создавать новый класс для страницы?

**Reply by lee-to** [reply]
Можно но через метод ресурса onLoad

```php
protected function onLoad(): void
{
$this->getIndexPage()->...
}

**Reply by ArtemYurov** [reply]
Да, работает (правда при таком отображение исчезла кнопка Сбросить, но это другой разговор)
```
    public function onLoad(): void
    {
        $this->getIndexPage()->pushToLayer(Layer::TOP, (new FiltersForm($this))());
    }
```

Но автодополнение не срабатывает,
<img width="762" alt="image" src="https://github.com/user-attachments/assets/de741be5-a681-49b5-981f-4accca41c717">

**Reply by lee-to** [reply]
с дженериками разберемся ближе к релизу, пока не до них

**Reply by ArtemYurov** [reply]
очень важная штука, как дополнение документации

**Comment by ArtemYurov** [comment]
Обновление с 2 до 3  (мой список действий)
[TEST_UPGRADE_2_to_3.md](https://github.com/user-attachments/files/16921575/TEST_UPGRADE_2_to_3.md)
Черновик от lee-to [UPGRADE.md](https://github.com/user-attachments/files/16921452/UPGRADE.md)

**Reply by ArtemYurov** [reply]
Вопрос, куда делась настройка лого  в конфиге?
```
'logo' => env('MOONSHINE_LOGO', '/images/logo.png'),
'logo_small' => env('MOONSHINE_LOGO_SMALL', '/images/logo.png'),
```

**Reply by lee-to** [reply]
Не имеет смысла, теперь делается через Layout, передается в компонент Logo

**Reply by ArtemYurov** [reply]
`app/MoonShine/Layouts/MoonShineLayout.php`
```
    protected function getLogoComponent(): Logo
    {
        return Logo::make(
            $this->getHomeUrl(),
            '/images/logo.png',
            '/images/logo.png'
        );
    }
```

**Reply by ArtemYurov** [reply]
в ресурсах поменять
```
    public function getActiveActions(): array
    {
        return ['view','update'];
    }
```
на 
```
    protected function activeActions(): ListOf
    {
        return parent::activeActions()->except(Action::DELETE, Action::MASS_DELETE);
    }
```
или
```
    protected function activeActions(): ListOf
    {
        return new ListOf(Action::class, []);
    }
```

**Reply by ArtemYurov** [reply]
Небольшой helper когда нужно скопировать поля из indexFields с добавлением\исключением в detail или from
```
<?php

namespace App\MoonShine\Classes;

use Illuminate\Support\Collection;

class FieldsHelper extends Collection
{

    public function afterColumn(string $column, array $fields): FieldsHelper
    {
        $items = [];
        foreach ($this->items as $item) {
            $items[] = $item;
            if ($item->getColumn() == $column) {
                foreach ($fields as $field)
                    $items[] = $field;
            }
        }

        $this->items = $items;

        return $this;
    }

    public function exceptColumns(string|array $columns): FieldsHelper
    {
        $items = [];
        foreach ($this->items as $item) {
            if (is_string($columns) ? $item->getColumn() != $columns : !in_array($item->getColumn(), $columns)) {
                $items[] = $item;
            }
        }

        $this->items = $items;

        return $this;
    }


}
```

Пример использования
```
    protected function detailFields(): iterable
    {
        return FieldsHelper::make($this->indexFields())
            ->afterColumn('expenses', [
                Json::make('Детально', 'expenses_detail')->keyValue()
            ])
            ->afterColumn('expenses_other', [
                Json::make('Детально', 'expenses_other_detail')->keyValue()
            ])
            ->exceptColumns('id');
    }
```

**Reply by lee-to** [reply]
> Небольшой helper когда нужно скопировать поля из indexFields с добавлением\исключением в detail или from
> 
> ```
> 
> <?php
> 
> 
> 
> namespace App\MoonShine\Classes;
> 
> 
> 
> use Illuminate\Support\Collection;
> 
> 
> 
> class FieldsHelper extends Collection
> 
> {
> 
> 
> 
>     public function afterColumn(string $column, array $fields): FieldsHelper
> 
>     {
> 
>         $items = [];
> 
>         foreach ($this->items as $item) {
> 
>             $items[] = $item;
> 
>             if ($item->getColumn() == $column) {
> 
>                 foreach ($fields as $field)
> 
>                     $items[] = $field;
> 
>             }
> 
>         }
> 
> 
> 
>         $this->items = $items;
> 
> 
> 
>         return $this;
> 
>     }
> 
> 
> 
>     public function exceptColumns(string|array $columns): FieldsHelper
> 
>     {
> 
>         $items = [];
> 
>         foreach ($this->items as $item) {
> 
>             if (is_string($columns) ? $item->getColumn() != $columns : !in_array($item->getColumn(), $columns)) {
> 
>                 $items[] = $item;
> 
>             }
> 
>         }
> 
> 
> 
>         $this->items = $items;
> 
> 
> 
>         return $this;
> 
>     }
> 
> 
> 
> 
> 
> }
> 
> ```
> 
> 
> 
> Пример использования
> 
> ```
> 
>     protected function detailFields(): iterable
> 
>     {
> 
>         return FieldsHelper::make($this->indexFields())
> 
>             ->afterColumn('expenses', [
> 
>                 Json::make('Детально', 'expenses_detail')->keyValue()
> 
>             ])
> 
>             ->afterColumn('expenses_other', [
> 
>                 Json::make('Детально', 'expenses_other_detail')->keyValue()
> 
>             ])
> 
>             ->exceptColumns('id');
> 
>     }
> 
> ```

Можете создать пакет с хелперами, будете его постоянно пополнять

**Comment by ArtemYurov** [comment]
Опять поле Position, если попробовать сделать Position::make()->sortable(), то при попытке отсортировать получается так:
`Column not found: 1054 Unknown column '#' in 'order clause' (Connection: mysql, SQL: select * from `profit_reports` order by `#` asc)`

**Reply by lee-to** [reply]
ну каким чудом он будет сортировать, это просто индекс, надо переопределять тогда билдер

**Reply by ArtemYurov** [reply]
мало ли, вдруг волшебник поработал там. просто случайно поставил сортироваться это поле

**Reply by lee-to** [reply]
можно добавить исключение или написать в доке alert для волшебников

**Comment by ArtemYurov** [comment]
https://github.com/moonshine-software/moonshine/blob/3.x/src/Laravel/src/Fields/Relationships/BelongsTo.php#L55
Тут ошибка, в getPageUrl не передается правильный индекс для ресурса (как получить индекс, я не понял)

ещё вопрос, как лучше тут оформить сахар, чтобы можно было по умолчанию задавать переход на detailPage? 

Добавить protected ?PageType $type = NULL;,  public function setPageType(PageType $type): void, public function getPageType(): ?PageType и добавить проверку в resolvePreview?

**Comment by ArtemYurov** [comment]
По поводу `TableBuilder` и функции `headRows`  - нет возможности получить или сгенерировать стоковый headRows. Ну и для footRows тоже

**Reply by lee-to** [reply]
> По поводу `TableBuilder` и функции `headRows`  - нет возможности получить или сгенерировать стоковый headRows. Ну и для footRows тоже

Через замыкания

**Reply by ArtemYurov** [reply]
окей
```
        $table->headRows(fn($default) => TableRows::make([$default]));
        $table->footRows(fn($default) => TableRows::make([$default]));
```

**Comment by ArtemYurov** [comment]


**Reply by lee-to** [reply]
> Такой финт ушами не прокатывает
> 
> ```
> 
> protected function components(): iterable
> 
> {
> 
>         return [
> 
>               HasMany::make('',resource: ProfitReportResource::class)
> 
>                   ->name('current-month')
> 
>                   ->modifyBuilder(fn($builder) =>
> 
>                           ProfitReport::query()->where('id',2)
> 
>                   ),
> 
>               Divider::make('Предыдущий месяц'),
> 
>               HasMany::make('',resource: ProfitReportResource::class)
> 
>                   ->name('previous-month')
> 
>                   ->modifyBuilder(fn($builder) =>
> 
>                       ProfitReport::query()->where('id',5)
> 
>                   ),
> 
>         ];
> 
> }
> 
> ```
> 
> 
> 
> сначала выполняется
> 
> `select * from `profit_reports` where `id` = 2 order by `date` desc`
> 
> а потом
> 
> `select * from `profit_reports` where `id` = 2 order by `date` desc, `date` desc`

Путаете modify с полной заменой, такая полная замена уже ничего общего с отношением hasmany иметь не будет

**Reply by lee-to** [reply]
Еще и в components))

**Reply by lee-to** [reply]
> > Еще и в components))
> 
> хотелось же как всегда попроще, получить готовую таблицу 

Если решили играть против правил, надо хотя бы знать правила и огромная просьба в этом топике помогать с багами а не фантазировать

**Reply by lee-to** [reply]
Да и просто сделать TableBuilder было бы проще чем выдумывать небылицы

**Comment by ArtemYurov** [comment]
Вопрос такой, name у компонента должен быть уникальным только на странице или во всём проекте?

Я сделал свою Page c TableBuilder и сделал несколько страниц через extends.
Использовал columnSelection и в итоге у меня общий выбор колонок всех страницах. В JS для storage я так понял используется name таблицы.
С одной стороны это может быть и правильно, но тогда  необходимо это упомятуть в документации.

**Reply by lee-to** [reply]
Да действительно! Добавил Warning в доку

**Comment by ArtemYurov** [comment]
когда на custom Page добавляю 
```
TableBuilder::make()
->buttons([
                ActionButton::make('Mass Delete')->bulk()
            ])
```
то не срабатывает actionsOpen для показа кнопок
<img width="1011" alt="image" src="https://github.com/user-attachments/assets/10da457c-502c-410d-9c24-4efe4aa4048c">

**Reply by lee-to** [reply]
Добавил фикс этого кейса и немного прокачал возможные проблемы при использовании таблиц без кастов, когда не ясен какой ключ использовать для операций https://github.com/moonshine-software/doc/blob/3.x/ru/components/table-builder.md#type-cast

**Comment by ArtemYurov** [comment]
дано
```
class TelegramUser extends Model
    public function telegraph_bot(): BelongsTo
   {
        return $this->belongsTo(TelegraphBot::class);
    }

    public function telegraph_bot_permissions(): BelongsToMany
    {
        return $this->belongsToMany(TelegraphBotPermission::class);
    }
```
```
class TelegraphBotPermission extends Model
{
    public function telegraph_bot(): BelongsTo
    {
        return $this->belongsTo(TelegraphBot::class);
    }

}
```

`TelegramUserResource`:

```
            BelongsTo::make('Бот', 'telegraph_bot', TelegraphBotResource::class)->sortable()->disabled(),
            BelongsToMany::make('Разрешения','telegraph_bot_permissions','name', resource: TelegraphBotPermissionResource::class)
                ->associatedWith('telegraph_bot_id'),
```

вот associatedWith не работает (в 2 работало)

без метода последний вызываемый query такой
`select * from telegraph_bot_permissions` и соответственно показывает все варианты 

а с ним (этот запрос ещё вызывался до этого в обоих случаях)
```
select
  `telegraph_bot_permissions`.*,
  `telegram_user_telegraph_bot_permission`.`telegram_user_id` as `pivot_telegram_user_id`,
  `telegram_user_telegraph_bot_permission`.`telegraph_bot_permission_id` as `pivot_telegraph_bot_permission_id`
from
  `telegraph_bot_permissions`
  inner join `telegram_user_telegraph_bot_permission` on `telegraph_bot_permissions`.`id` = `telegram_user_telegraph_bot_permission`.`telegraph_bot_permission_id`
where
  `telegram_user_telegraph_bot_permission`.`telegram_user_id` = 10
```
